src/v/cluster/controller_snapshot.cc

```diff
@@ -13,11 +13,78 @@
 
 namespace cluster {
 
+namespace controller_snapshot_parts {
+
+template<typename Vec>
+ss::future<> write_vector_async(iobuf& out, Vec t) {
+    if (unlikely(t.size() > std::numeric_limits<serde::serde_size_t>::max())) {
+        throw serde::serde_exception(fmt_with_ctx(
+          ssx::sformat,
+          "serde: vector size {} exceeds serde_size_t",
+          t.size()));
+    }
+    serde::write(out, static_cast<serde::serde_size_t>(t.size()));
+    return ss::do_with(std::move(t), [&out](Vec& t) {
+        return ss::do_for_each(
+          t, [&out](auto& el) { serde::write(out, std::move(el)); });
+    });
+}
+
+template<typename T>
+concept Reservable = requires(T t) {
+    t.reserve(0U);
+};
+
+template<typename Vec>
+ss::future<Vec>
+read_vector_async_nested(iobuf_parser& in, std::size_t const bytes_left_limit) {
+    using value_type = typename Vec::value_type;
+    const auto size = serde::read_nested<serde::serde_size_t>(
+      in, bytes_left_limit);
+    return ss::do_with(Vec{}, [size, &in, bytes_left_limit](Vec& t) {
+        if constexpr (Reservable<Vec>) {
+            t.reserve(size);
+        }
```

Comment:
neato

Reply:
Lifted from the Felix's tag-invoke pr. This probably should go into the main serde code after it is merged.